launch_ros/launch_ros/utilities/evaluate_parameters.py

```diff
@@ -52,12 +54,29 @@ def evaluate_parameter_dict(
             if isinstance(value[0], Substitution):
                 # Value is a list of substitutions, so perform them to make a string
                 evaluated_value = perform_substitutions(context, list(value))
+                evaluated_value = yaml.safe_load(evaluated_value)
             elif isinstance(value[0], Sequence):
                 # Value is an array of a list of substitutions
                 output_subvalue = []  # List[str]
                 for subvalue in value:
-                    output_subvalue.append(perform_substitutions(context, list(subvalue)))
+                    value = perform_substitutions(context, list(subvalue))
+                    output_subvalue.append(value)
                     evaluated_value = tuple(output_subvalue)
+                # All values in a list must have the same type.
+                # If they don't then assume it is a list of strings
+                yaml_evaluated_value = []  # Union[List[str], List[int], List[float], List[bool]]
```

Comment:
nitpick: now that we don't have to support python3.5 we can use the normal syntax for this:

```python
yaml_evaluated_value: Union[List[str], List[int], List[float], List[bool]] = []
```

Reply:
Replaced comments with new syntax in this function in 379e43f